independent-projects/tools/common/src/main/java/io/quarkus/devtools/commands/handlers/QuarkusCommandHandlers.java

```diff
@@ -11,111 +16,86 @@
+                builder.add(Extensions.parse(query));
+            } else {
+                SelectionResult result = select(invocation.getPlatformDescriptor().getExtensions(), query, false);
+                if (!result.matches()) {
+                    StringBuilder sb = new StringBuilder();
+                    // We have 3 cases, we can still have a single candidate, but the match is on label
+                    // or we have several candidates, or none
+                    Set<Extension> candidates = result.getExtensions();
+                    if (candidates.isEmpty()) {
+                        // No matches at all.
+                        invocation.log().info(nok(" Cannot find a dependency matching '" + query + "', maybe a typo?"));
+                        return null;
+                    } else {
+                        sb.append(NOK).append(" Multiple extensions matching '").append(query).append("'");
+                        result.getExtensions()
+                                .forEach(extension -> sb.append(System.lineSeparator()).append("     * ")
+                                        .append(extension.managementKey()));
+                        sb.append(System.lineSeparator())
+                                .append("     Be more specific e.g using the exact name or the full GAV.");
+                        invocation.log().info(sb.toString());
+                        return null;
+                    }
+                } else { // Matches.
+                    for (Extension extension : result) {
+                        // Don't set success to false even if the dependency is not added; as it's should be idempotent.
+                        builder.add(extension);
                     }
                 }
             }
-        } catch (IOException e) {
-            throw new QuarkusCommandException("Failed to add extensions", e);
-        }
-
-        if (buildFile != null && updated) {
-            try {
-                buildFile.close();
-            } catch (IOException e) {
-                throw new QuarkusCommandException("Failed to update the project", e);
-            }
         }
-
-        return new QuarkusCommandOutcome(success).setValue(AddExtensions.OUTCOME_UPDATED, updated);
+        return builder.build();
     }
 
     /**
      * Selection algorithm.
      *
+     * @param allPlatformExtensions the list of all platform extensions
      * @param query the query
-     * @param extensions the extension list
      * @param labelLookup whether or not the query must be tested against the labels of the extensions. Should
      *        be {@code false} by default.
      * @return the list of matching candidates and whether or not a match has been found.
      */
-    static SelectionResult select(String query, List<Extension> extensions, boolean labelLookup) {
+    static SelectionResult select(final List<Extension> allPlatformExtensions, final String query, final boolean labelLookup) {
```

Comment:
why the change of signature here ?

Reply:
TBH I don't remember 😅 the name is clearer, but the order 🤗, I'll switch the order back..

Reply:
fixed